project-koku/koku
# (1639) add celery task to clean worker volume weekly

This PR adds a weekly celery task (scheduled to run on Sunday at midnight) to delete any files on the volume that do not belong to manifests that we need to process. 

Testing:
--------
1. Remove the `if not settings.DEVELOPMENT` condition and edit the schedule for the `masu.celery.tasks.clean_volume` task in `koku/koku/celery.py` to run every minute: 
```
app.conf.beat_schedule["clean_volume"] = {
    "task": "masu.celery.tasks.clean_volume",
    "schedule": crontab(minute="*")
}
```
2. Bring up the project as normal and look at the koku-worker logs: 
```
make docker-up
docker-compose -f logs koku-worker
```
3. Populate the manifest table: 
first make sure that you have the following in your env file 
```
API_PATH_PREFIX='/api/cost-management'
NISE_REPO_PATH=/path/to/nise
KOKU_API_HOSTNAME=127.0.0.1
KOKU_PORT=8000
MASU_API_HOSTNAME=127.0.0.1
MASU_PORT=5000
```
and then run the following: 
```
make docker-reinitdb-with-providers
make load-test-customer-data
```
4. Now exec into your db and view all of the records for the manifest table:
```
docker exec -it koku_db psql -U postgres
SELECT * FROM reporting_common_costusagereportmanifest;
```
5. Now we are going to insert a fake record into the manifest table that has not finished processing - however modify the below to have an existing provider id - (replace `9be19863-cbfc-472d-bab2-b47ddd93c8ac` with your valid provider_id) : 
```
INSERT INTO reporting_common_costusagereportmanifest(assembly_id, manifest_creation_datetime, manifest_updated_datetime, num_processed_files, num_total_files, provider_id, billing_period_start_datetime) VALUES ('234567', '2020-02-19 22:09:20.202402+00', '2020-02-19 22:10:04.688277+00', 0, 1, '9be19863-cbfc-472d-bab2-b47ddd93c8ac', '2020-02-01 00:00:00+00');
```
Now - relook at the entries for the reporting_common_costusagereportmanifest table and make sure that your new entry is there. 

6. Create some random files in your PVC_DIR - this is in `koku/testing/pvc_dir/` if you are running using docker - you can create files at the top level or in subdirectories. Also create a file that is prefixed with the assembly_id from the entry that you created - (`234567` if you copied my insert): 
```
(koku-iU2DMmAw) bash-3.2$ pwd
/Users/ashleybaiken/cmdevelopment/koku/testing/pvc_dir
(koku-iU2DMmAw) bash-3.2$ touch 234567.csv
(koku-iU2DMmAw) bash-3.2$ touch fakedata.csv
```
7. Finally check the koku-worker logs - you should see something like the following: 
```
koku_worker       | [2020-02-19 14:49:00,046: INFO/ForkPoolWorker-1] [masu.celery.tasks.clean_volume(636d25dc-470d-4540-8a9e-1ebcbed01344 via 636d25dc-470d-4540-8a9e-1ebcbed01344)] Cleaning up the volume at /testing/pvc_dir 
koku_worker       | [2020-02-19 14:49:00,101: INFO/ForkPoolWorker-1] [masu.celery.tasks.clean_volume(636d25dc-470d-4540-8a9e-1ebcbed01344 via 636d25dc-470d-4540-8a9e-1ebcbed01344)] The following files were deleted: 
koku_worker       | [2020-02-19 14:49:00,101: INFO/ForkPoolWorker-1] [masu.celery.tasks.clean_volume(636d25dc-470d-4540-8a9e-1ebcbed01344 via 636d25dc-470d-4540-8a9e-1ebcbed01344)] ['/testing/pvc_dir/fakerdata.csv', '/testing/pvc_dir/fakerr/fakest.csv']
```
8. Check to make sure that the file you created with the assembly_id prefix is still there and that the other files have been discarded.
